File: stock_search/etf.py
"""Resolve ETF holdings and sector snapshots with cache support."""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from datetime import datetime, timedelta
import re
from typing import Any

from stock_search.common_utils import normalize_ticker_symbol
from stock_search.data_sources.stockanalysis import StockAnalysisSource
from stock_search.data_sources.yahoofinance import ETF_QUOTE_TYPE, YahooFinanceSource
from stock_search.models import SECTOR_LABELS, SECTOR_PATTERN_RULES, ETFSector, Holding

ETF_CACHE_MAX_AGE_DAYS = 7

SECTOR_REGEX_MAP: tuple[tuple[re.Pattern[str], str], ...] = tuple((re.compile(pattern, flags=re.IGNORECASE), label) for pattern, label in SECTOR_PATTERN_RULES)


@dataclass(frozen=True)
class ETFSnapshotResult:
    """Represent one resolved ETF snapshot result."""

    holdings: list[Holding]
    sectors: list[ETFSector]
    error: str | None


@dataclass(frozen=True)
class ETFResolutionResult:
    """Represent ETF resolution results for one portfolio pass."""

    stock_positions: list[dict[str, Any]]
    etf_positions: list[dict[str, Any]]
    snapshot_by_ticker: dict[str, ETFSnapshotResult]
    etf_refreshed_count: int
    cache_changed: bool


def _pool_workers(item_count: int, cap: int | None = None) -> int:
    """Return the thread-pool size used for ETF fetches."""
    if item_count <= 0:
        return 1
    if cap is None:
        return item_count
    return max(1, min(item_count, cap))


def _is_cache_entry_fresh(fetched_at: str, *, now: datetime) -> bool:
    """Return whether an ETF cache entry is still fresh."""
    try:
        fetched_at_dt = datetime.fromisoformat(fetched_at)
    except ValueError:
        return False
    return now - fetched_at_dt <= timedelta(days=ETF_CACHE_MAX_AGE_DAYS)


def normalize_sector_name(raw: str | None) -> str:
    """Normalize sector labels into a stable display form."""
    sector_text = (raw or "").strip()
    if not sector_text:
        return SECTOR_LABELS["other"]
    for pattern, canonical_label in SECTOR_REGEX_MAP:
        if pattern.search(sector_text):
            return canonical_label
    return SECTOR_LABELS["other"]


def _parse_cached_snapshot(
    stats_data: dict[str, Any],
    *,
    ticker: str,
    now: datetime,
    require_fresh: bool,
) -> tuple[list[Holding], list[ETFSector]] | None:
    """Parse cached snapshot."""
    entry = stats_data.get(ticker, {})
    if not isinstance(entry, dict):
        return None

    fetched_at = entry.get("etf_holdings_fetched_at")
    holdings_raw = entry.get("etf_holdings")
    sectors_raw = entry.get("etf_sectors")
    if not isinstance(fetched_at, str) or not isinstance(holdings_raw, list):
        return None
    if require_fresh and not _is_cache_entry_fresh(fetched_at, now=now):
        return None

    try:
        holdings = [Holding.model_validate(holding) for holding in holdings_raw]
    except Exception:
        return None

    sectors: list[ETFSector] = []
    if isinstance(sectors_raw, list):
        for sector in sectors_raw:
            if not isinstance(sector, dict):
                continue
            name = normalize_sector_name(str(sector.get("name") or "").strip())
            weight_raw = sector.get("weight")
            if not isinstance(weight_raw, (int, float)):
                continue
            sectors.append(ETFSector(name=name, weight=float(weight_raw)))

    return holdings, sectors


def load_etf_cache_from_stats(stats_data: dict[str, Any], ticker: str, now: datetime) -> tuple[list[Holding], list[ETFSector]] | None:
    """Load ETF cache from stats."""
    return _parse_cached_snapshot(
        stats_data,
        ticker=normalize_ticker_symbol(ticker),
        now=now,
        require_fresh=True,
    )


def _load_stale_cache_from_stats(
    stats_data: dict[str, Any],
    ticker: str,
    now: datetime,
) -> tuple[list[Holding], list[ETFSector]] | None:
    """Load stale ETF cache entries from the stats store."""
    return _parse_cached_snapshot(
        stats_data,
        ticker=normalize_ticker_symbol(ticker),
        now=now,
        require_fresh=False,
    )


def _sector_rows(snapshot_sectors: object) -> list[ETFSector]:
    """Convert sector weights into normalized table rows."""
    if not hasattr(snapshot_sectors, "model_dump"):
        return []
    rows: list[ETFSector] = []
    for key, value in snapshot_sectors.model_dump().items():
        if value is None:
            continue
        sector_name = SECTOR_LABELS.get(str(key), str(key).replace("_", " ").title())
        rows.append(ETFSector(name=normalize_sector_name(sector_name), weight=float(value)))
    rows.sort(key=lambda row: row.weight, reverse=True)
    return rows


def _serialized_cache_payload(*, holdings: list[Holding], sectors: list[ETFSector], fetched_at: str) -> dict[str, Any]:
    """Serialize an ETF snapshot for cache storage."""
    return {
        "etf_holdings": [holding.model_dump() for holding in holdings],
        "etf_sectors": [sector.model_dump() for sector in sectors],
        "etf_holdings_fetched_at": fetched_at,
    }


def store_etf_cache_in_stats(
    stats_data: dict[str, Any],
    ticker: str,
    holdings: list[Holding],
    sectors: list[ETFSector],
    now: datetime,
) -> bool:
    """Store ETF cache in stats."""
    ticker_key = normalize_ticker_symbol(ticker)
    entry = stats_data.get(ticker_key, {})
    if not isinstance(entry, dict):
        entry = {}

    payload = _serialized_cache_payload(holdings=holdings, sectors=sectors, fetched_at=now.isoformat())
    changed = (
        entry.get("etf_holdings") != payload["etf_holdings"]
        or entry.get("etf_sectors") != payload["etf_sectors"]
        or entry.get("etf_holdings_fetched_at") != payload["etf_holdings_fetched_at"]
    )
    if changed:
        entry.update(payload)
        stats_data[ticker_key] = entry
    return changed


def is_etf_ticker(ticker: str) -> bool:
    """Return whether ETF ticker."""
    try:
        quote_type = YahooFinanceSource(ticker).get_quote_type()
    except Exception:
        return False
    return quote_type == ETF_QUOTE_TYPE


def get_etf_snapshot(ticker: str) -> tuple[list[Holding], list[ETFSector], str | None]:
    """Fetch one ETF snapshot from the available data sources."""
    try:
        snapshot = StockAnalysisSource(ticker).get_etf_holdings_snapshot()
    except Exception as exc:
        return [], [], str(exc)
    if snapshot is None:
        return [], [], "no snapshot returned"
    return snapshot.holdings.holdings, _sector_rows(snapshot.sectors), None


def _fetch_quote_type(ticker: str) -> tuple[str, bool]:
    """Fetch quote type."""
    return ticker, is_etf_ticker(ticker)


def _fetch_snapshot(ticker: str) -> tuple[str, list[Holding], list[ETFSector], str | None]:
    """Fetch snapshot."""
    holdings, sectors, error = get_etf_snapshot(ticker)
    return ticker, holdings, sectors, error


def _is_etf_from_stats(stats_data: dict[str, Any], ticker: str) -> bool | None:
    """Return whether ETF from stats."""
    entry = stats_data.get(ticker)
    if not isinstance(entry, dict):
        return None
    quote_type = str(entry.get("quote_type") or "").upper().strip()
    if not quote_type:
        return None
    return quote_type == ETF_QUOTE_TYPE


def classify_and_resolve_etfs(
    positions: list[dict[str, Any]],
    stats_data: dict[str, Any],
    now: datetime,
) -> ETFResolutionResult:
    """Split portfolio tickers into ETF and non-ETF groups and resolve snapshots."""
    stock_positions: list[dict[str, Any]] = []
    etf_positions: list[dict[str, Any]] = []
    snapshot_by_ticker: dict[str, ETFSnapshotResult] = {}
    unresolved_tickers: list[str] = []
    position_by_ticker = {normalize_ticker_symbol(position["ticker"]): position for position in positions}

    for position in positions:
        ticker = normalize_ticker_symbol(position["ticker"])
        cached_snapshot = load_etf_cache_from_stats(stats_data, ticker, now)
        if cached_snapshot is not None and cached_snapshot[1]:
            holdings, sectors = cached_snapshot
            etf_positions.append(position)
            snapshot_by_ticker[ticker] = ETFSnapshotResult(holdings=holdings, sectors=sectors, error=None)
            continue

        is_etf_from_stats = _is_etf_from_stats(stats_data, ticker)
        if is_etf_from_stats is True:
            etf_positions.append(position)
            continue
        if is_etf_from_stats is False:
            stock_positions.append(position)
            continue
        unresolved_tickers.append(ticker)

    if unresolved_tickers:
        with ThreadPoolExecutor(max_workers=_pool_workers(len(unresolved_tickers), cap=32)) as executor:
            classified = list(executor.map(_fetch_quote_type, unresolved_tickers))
        for ticker, is_etf in classified:
            pos = position_by_ticker[ticker]
            if is_etf:
                etf_positions.append(pos)
            else:
                stock_positions.append(pos)

    etf_tickers = {normalize_ticker_symbol(position["ticker"]) for position in etf_positions}
    for position in positions:
        ticker = normalize_ticker_symbol(position["ticker"])
        if ticker not in etf_tickers and position not in stock_positions:
            stock_positions.append(position)

    missing_tickers = [normalize_ticker_symbol(position["ticker"]) for position in etf_positions if normalize_ticker_symbol(position["ticker"]) not in snapshot_by_ticker]
    etf_refreshed_count = 0
    cache_changed = False
    if missing_tickers:
        with ThreadPoolExecutor(max_workers=_pool_workers(len(missing_tickers), cap=16)) as executor:
            fetched = list(executor.map(_fetch_snapshot, missing_tickers))
        for ticker, holdings, sectors, error in fetched:
            if error is None:
                etf_refreshed_count += 1
                cache_changed = store_etf_cache_in_stats(stats_data, ticker, holdings, sectors, now) or cache_changed
                snapshot_by_ticker[ticker] = ETFSnapshotResult(holdings=holdings, sectors=sectors, error=None)
                continue

            stale_snapshot = _load_stale_cache_from_stats(stats_data, ticker, now)
            if stale_snapshot is not None:
                stale_holdings, stale_sectors = stale_snapshot
                snapshot_by_ticker[ticker] = ETFSnapshotResult(
                    holdings=stale_holdings,
                    sectors=stale_sectors,
                    error=error,
                )
            else:
                snapshot_by_ticker[ticker] = ETFSnapshotResult(holdings=[], sectors=[], error=error)

    return ETFResolutionResult(
        stock_positions=stock_positions,
        etf_positions=etf_positions,
        snapshot_by_ticker=snapshot_by_ticker,
        etf_refreshed_count=etf_refreshed_count,
        cache_changed=cache_changed,
    )

File: src/stock-search/etf.ts
/** Resolve ETF holdings and sector snapshots with cache support. */

import {
	StockAnalysisSource,
	type StockAnalysisEtfSnapshot,
} from "./data-sources/stockanalysis/index.js";
import { ETF_QUOTE_TYPE, YahooFinanceSource } from "./data-sources/yahoo-finance.js";
import type { BackendStore, PositionRow, StockEntry } from "./api/data-store.js";
import { SECTOR_LABELS, SECTOR_PATTERN_RULES } from "./models/labels.js";
import { normalizeTicker } from "./utils.js";

export type EtfHolding = {
	ticker: string;
	name: string | null;
	weight: number;
};

export type EtfSector = {
	name: string;
	weight: number;
};

export type EtfSnapshotResult = {
	holdings: EtfHolding[];
	sectors: EtfSector[];
	error: string | null;
};

export type EtfResolutionResult = {
	stockPositions: PositionRow[];
	etfPositions: PositionRow[];
	snapshotByTicker: Record<string, EtfSnapshotResult>;
	etfRefreshedCount: number;
	cacheChanged: boolean;
	changedTickers: string[];
};

export type EtfSnapshotCacheResult = {
	snapshot: EtfSnapshotResult;
	refreshedIndicators: Record<string, unknown> | null;
	didRefresh: boolean;
};

const ETF_CACHE_MAX_AGE_MS = 7 * 24 * 60 * 60 * 1000;
function parseTimestamp(value: unknown): number | null {
	if (typeof value !== "string" || !value.trim()) {
		return null;
	}
	const timestamp = Date.parse(value);
	return Number.isFinite(timestamp) ? timestamp : null;
}

/** Normalize sector labels into the stable display values used by the app. */
export function normalizeSectorName(value: string | null | undefined): string {
	const sectorText = String(value ?? "").trim();
	if (!sectorText) {
		return SECTOR_LABELS.other;
	}
	for (const label of Object.values(SECTOR_LABELS)) {
		if (sectorText.toLowerCase() === label.toLowerCase()) {
			return label;
		}
	}
	for (const [pattern, label] of SECTOR_PATTERN_RULES) {
		if (new RegExp(pattern, "i").test(sectorText)) {
			return label;
		}
	}
	return SECTOR_LABELS.other;
}

function loadEtfCache(
	indicators: Record<string, unknown>,
	now: number,
	requireFresh: boolean,
): { holdings: EtfHolding[]; sectors: EtfSector[] } | null {
	const fetchedAt = parseTimestamp(indicators.etf_holdings_fetched_at);
	if (requireFresh && (fetchedAt == null || fetchedAt < now - ETF_CACHE_MAX_AGE_MS)) {
		return null;
	}
	if (!Array.isArray(indicators.etf_holdings)) {
		return null;
	}

	const holdings = indicators.etf_holdings
		.filter((holding) => typeof holding === "object" && holding !== null)
		.map((holding) => holding as Record<string, unknown>)
		.map((holding) => ({
			ticker: normalizeTicker(holding.ticker),
			name:
				typeof holding.name === "string" && holding.name.trim()
					? holding.name.trim()
					: null,
			weight: Number(holding.weight),
		}))
		.filter((holding) => holding.ticker && Number.isFinite(holding.weight));
	const sectors = Array.isArray(indicators.etf_sectors)
		? normalizeEtfSectors(
				indicators.etf_sectors
					.filter((sector) => typeof sector === "object" && sector !== null)
					.map((sector) => sector as Record<string, unknown>)
					.map((sector) => ({
						name: typeof sector.name === "string" ? sector.name : null,
						weight: Number(sector.weight),
					})),
			)
		: [];

	return { holdings, sectors };
}

function emptyEtfSnapshot(error: string | null = null): EtfSnapshotResult {
	return {
		holdings: [],
		sectors: [],
		error,
	};
}

function normalizeEtfSectors(
	sectors: Array<{ name: string | null | undefined; weight: number }>,
): EtfSector[] {
	const weightsBySector = new Map<string, number>();
	for (const sector of sectors) {
		if (!Number.isFinite(sector.weight)) {
			continue;
		}
		const name = normalizeSectorName(sector.name);
		weightsBySector.set(name, (weightsBySector.get(name) ?? 0) + sector.weight);
	}
	return [...weightsBySector.entries()].map(([name, weight]) => ({
		name,
		weight: Number(weight.toFixed(4)),
	}));
}

function storeEtfCache(
	indicators: Record<string, unknown>,
	holdings: EtfHolding[],
	sectors: EtfSector[],
	now: number,
): Record<string, unknown> {
	return {
		...indicators,
		etf_holdings: holdings,
		etf_sectors: sectors,
		etf_holdings_fetched_at: new Date(now).toISOString(),
	};
}

/** Resolve one ETF snapshot with cache-first semantics and optional live refresh. */
export async function resolveEtfSnapshotCache(
	tickerInput: string,
	stockEntry: StockEntry | null,
	allowLiveFetch: boolean,
	now = Date.now(),
): Promise<EtfSnapshotCacheResult> {
	const ticker = normalizeTicker(tickerInput);
	const indicators = stockEntry?.indicators ?? {};
	const cachedSnapshot = loadEtfCache(indicators, now, true);
	if (cachedSnapshot) {
		return {
			snapshot: {
				holdings: cachedSnapshot.holdings,
				sectors: cachedSnapshot.sectors,
				error: null,
			},
			refreshedIndicators: null,
			didRefresh: false,
		};
	}

	const staleSnapshot = loadEtfCache(indicators, now, false);
	if (!allowLiveFetch) {
		return {
			snapshot: staleSnapshot
				? {
						holdings: staleSnapshot.holdings,
						sectors: staleSnapshot.sectors,
						error: null,
					}
				: emptyEtfSnapshot(),
			refreshedIndicators: null,
			didRefresh: false,
		};
	}

	const snapshot = await getEtfSnapshot(ticker);
	if (!snapshot.error) {
		return {
			snapshot,
			refreshedIndicators: storeEtfCache(
				{ ...indicators, quote_type: ETF_QUOTE_TYPE },
				snapshot.holdings,
				snapshot.sectors,
				now,
			),
			didRefresh: true,
		};
	}

	return {
		snapshot: staleSnapshot
			? {
					holdings: staleSnapshot.holdings,
					sectors: staleSnapshot.sectors,
					error: snapshot.error,
				}
			: emptyEtfSnapshot(snapshot.error),
		refreshedIndicators: null,
		didRefresh: false,
	};
}

/** Fetch one ETF holdings snapshot from StockAnalysis. */
export async function getEtfSnapshot(
	tickerInput: string,
): Promise<{ holdings: EtfHolding[]; sectors: EtfSector[]; error: string | null }> {
	const ticker = normalizeTicker(tickerInput);
	const snapshot: StockAnalysisEtfSnapshot =
		await new StockAnalysisSource(ticker).getEtfHoldingsSnapshot();
	return {
		holdings: snapshot.holdings,
		sectors: normalizeEtfSectors(snapshot.sectors),
		error: snapshot.error,
	};
}

async function isEtfTicker(
	ticker: string,
	stockEntry: StockEntry | null,
): Promise<boolean> {
	const cachedQuoteType = String(stockEntry?.indicators.quote_type ?? "")
		.trim()
		.toUpperCase();
	if (cachedQuoteType) {
		return cachedQuoteType === ETF_QUOTE_TYPE;
	}
	const liveIndicators = await new YahooFinanceSource(ticker).getIndicatorsSnapshot();
	return String(liveIndicators.quote_type ?? "").toUpperCase() === ETF_QUOTE_TYPE;
}

/** Split held positions into ETF and stock groups and resolve ETF snapshots. */
export async function classifyAndResolveEtfs(
	store: BackendStore,
	positions: PositionRow[],
	stockMap: Record<string, StockEntry>,
	allowLiveFetch: boolean,
): Promise<EtfResolutionResult> {
	const stockPositions: PositionRow[] = [];
	const etfPositions: PositionRow[] = [];
	const snapshotByTicker: Record<string, EtfSnapshotResult> = {};
	const now = Date.now();

	const resolvedPositions = await Promise.all(
		positions.map(async (position) => {
			const ticker = normalizeTicker(position.ticker);
			if (!ticker || Number(position.quantity ?? 0) <= 0) {
				return null;
			}

			const stockEntry = stockMap[ticker] ?? null;
			const indicators = stockEntry?.indicators ?? {};
			const etfLike =
				String(indicators.quote_type ?? "").trim().toUpperCase() === ETF_QUOTE_TYPE ||
				(loadEtfCache(indicators, now, false)?.holdings.length ?? 0) > 0 ||
				(allowLiveFetch && (await isEtfTicker(ticker, stockEntry)));

			if (!etfLike) {
				return {
					kind: "stock" as const,
					position,
					ticker,
					stockEntry,
					indicators,
				};
			}

			const snapshotCache = await resolveEtfSnapshotCache(
				ticker,
				stockEntry,
				allowLiveFetch,
				now,
			);

			return {
				kind: "etf" as const,
				position,
				ticker,
				stockEntry,
				indicators,
				snapshot: snapshotCache.snapshot,
				refreshedIndicators: snapshotCache.refreshedIndicators,
				didRefresh: snapshotCache.didRefresh,
			};
		}),
	);

	const upserts: Array<{
		ticker: string;
		indicators?: Record<string, unknown>;
		evaluation?: Record<string, unknown>;
		labels?: string[];
	}> = [];
	const changedTickers = new Set<string>();
	let etfRefreshedCount = 0;

	for (const resolved of resolvedPositions) {
		if (!resolved) {
			continue;
		}

		if (resolved.kind === "stock") {
			stockPositions.push(resolved.position);
			if (
				String(resolved.indicators.quote_type ?? "").trim().toUpperCase() !== "EQUITY" &&
				allowLiveFetch
			) {
				upserts.push({
					ticker: resolved.ticker,
					indicators: { ...resolved.indicators, quote_type: "EQUITY" },
					evaluation: resolved.stockEntry?.evaluation ?? {},
					labels: resolved.stockEntry?.labels ?? [],
				});
			}
			continue;
		}

		etfPositions.push(resolved.position);
		snapshotByTicker[resolved.ticker] = resolved.snapshot;

		if (!resolved.didRefresh || !resolved.refreshedIndicators) {
			continue;
		}

		etfRefreshedCount += 1;
		changedTickers.add(resolved.ticker);
		upserts.push({
			ticker: resolved.ticker,
			indicators: resolved.refreshedIndicators,
			evaluation: resolved.stockEntry?.evaluation ?? {},
			labels: resolved.stockEntry?.labels ?? [],
		});
	}

	if (upserts.length > 0) {
		await store.upsertStocks(upserts);
	}

	return {
		stockPositions,
		etfPositions,
		snapshotByTicker,
		etfRefreshedCount,
		cacheChanged: upserts.length > 0,
		changedTickers: [...changedTickers],
	};
}
